### PR TITLE
Added A-Frame typings

### LINKED ISSUE #11696: [WIP] A-Frame typings
I have started working on TS 2 typings for [A-Frame](https://aframe.io). I figured I should file a ticket to avoid any duplication of work and (potentially) help out anyone looking for A-Frame typings in the meantime. They're a WIP published here https://github.com/devpaul/aframe-typings.

I'm using a testing library ATM to ensure all of my assumptions are correct about A-Frame, especially in regards to optional types, but I'm not really sure how to boil that down to code that might exercise the API without rewriting things outside of the testing framework. Advice is welcome.

### PR DESCRIPTION
**Pull request by devpaul**
Resolves #11696

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

File: types/aframe/tslint.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "extends": "dtslint/dt.json",
+  "rules": {
```
Comment (mhegazy) [inline types/aframe/tslint.json]: remove empty `rules`.

File: types/aframe/tests/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (mhegazy) [inline types/aframe/tests/tsconfig.json]: please remove this file

File: types/aframe/tests/tests.html [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE html>
```
Comment (mhegazy) [inline types/aframe/tests/tests.html]: please remove this file

File: types/aframe/tests/package.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (mhegazy) [inline types/aframe/tests/package.json]: please remove this file

File: types/aframe/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    "sourceMap": true,
+    "strictNullChecks": true,
+    "noEmit": true,
+    "forceConsistentCasingInFileNames": true,
+    "typeRoots": [ "../" ],
+    "types": [ ]
+  },
+  "files": [
+    "index.d.ts",
+    "aframe-tests.ts"
```
Comment (mhegazy) [inline types/aframe/tsconfig.json]: Either include your `tests` folder here or remove them all together, this way the CI would verify that changes to the declaration file does not break the tests.

File: types/aframe/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
+{
+  "version": "2.3.2",
```
Comment (mhegazy) [inline types/aframe/tsconfig.json]: please remove the `"version"` tag

File: types/aframe/UNUSED_FILES.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+tests
```
Comment (mhegazy) [inline types/aframe/UNUSED_FILES.txt]: please remove this file

File: types/aframe/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
+# aframe typings
```
Comment (mhegazy) [inline types/aframe/README.md]: please remove this file

**Review by mhegazy — CHANGES_REQUESTED** [review]
Only definition tests should be added. please remove the other files.

File: types/aframe/tests/package.json (hunk 4 above)
Comment (devpaul) [inline types/aframe/tests/package.json]: The tests have a dependency on the intern test framework. Do you have a suggestion on how I can provide that dependency so people can run the tests?

Comment (devpaul) [inline types/aframe/tests/package.json]: I might be able to run the tests w/ `tests.html` and load the test framework via unpkg.org so `package.json` would not be necessary. The tests are important to help identify any changes in aframe as new versions are released. Does DT have any standards for providing tests that I could follow?

Comment (mhegazy) [inline types/aframe/tests/package.json]: The only tests the CI run are definition tests. no functional tests. the definition tests assert that the change to the declaration file does not cause errors in valid usage of the API.

Comment (devpaul) [inline types/aframe/tests/package.json]: That's understandable. I can make the rest of the changes you've requested, but I'd like to be able to keep these integration tests to help me (and other devs) track changes to aframe's API. The definition tests are OK, but it won't inform us if there's a change to the API between versions. I understand the tests wont be run by the CI.

File: types/aframe/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+		registerComponent(name: string, component: ComponentDefinition): ComponentConstructor;
+		registerElement(name: string, element: ANode): void;
+		registerGeometry(name: string, geometery: THREE.Geometry): Geometry;
+		registerPrimitive(name: string, primitive: PrimitiveDefinition): void;
+		registerShader(name: string, shader: any): void;
+		registerSystem(name: string, definition: SystemDefinition): void;
+		schema: SchemaUtils;
+		shaders: { [ key: string ]: ShaderDescriptor };
+		systems: { [key: string]: System };
+		THREE: any; // TODO THREE
```
Comment (hinell) [inline types/aframe/index.d.ts]: I think the ``THREE`` is missing here.

File: types/aframe/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+		tick?(time: number, timeDelta: number): void;
+		update(oldData: any): void;
+		updateSchema?(): void;
+
+		extendSchema(update: Schema): void;
+		flushToDOM(): void;
+	}
+
+	interface ComponentConstructor {
+		new (el: Entity, name: string, id: string): Component;
```
Comment (hinell) [inline types/aframe/index.d.ts]: Are you sure that this Isn't a [mistake](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes)?

**Comment by hinell** [comment]
Also it would be good if you add some major docs links 😉

File: types/aframe/tests/package.json (hunk 4 above)
Comment (mhegazy) [inline types/aframe/tests/package.json]: you can always host these into a different repo

File: types/aframe/index.d.ts (hunk 9 above)
Comment (devpaul) [inline types/aframe/index.d.ts]: THREE exports its library as a namespace so I wasn't sure how to declare that type here

File: types/aframe/index.d.ts (hunk 10 above)
Comment (devpaul) [inline types/aframe/index.d.ts]: It is technically a constructor function, i.e. an object that is newable, not a class. I'm on the fence about casting this as a class.

File: types/aframe/index.d.ts (hunk 9 above)
Comment (hinell) [inline types/aframe/index.d.ts]: Actually it is exported as global variable. Read this for more [info](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/release%20notes/TypeScript%202.0.md#support-for-umd-module-definitions)

Comment (devpaul) [inline types/aframe/index.d.ts]: It took some time to figure out, but if I use `typeof THREE` it seems to work. I'll update.

File: types/aframe/tests/package.json (hunk 4 above)
Comment (devpaul) [inline types/aframe/tests/package.json]: Yeah, if DT doesn't include tests in it's CI then a separate repo is almost as good as including code that may not be run by an engineer editing these types. I'll spin off this material in a separate repo and add comments to `index.d.ts`

**Comment by minestarks** [comment]
@devpaul please make the requested changes soon to avoid the PR being closed for inactivity. Thank you!

File: types/aframe/tsconfig.json (hunk 5 above)
Comment (devpaul) [inline types/aframe/tsconfig.json]: I've chosen to remove tests. I'll set up a separate repo and run tests via Travis on a cron. Best solution I can think of given the limitations.

**Pull request merged**